File: android/core/src/main/java/ru/maksimbulva/chess/core/engine/move/generator/RayMoveGenerator.kt
package ru.maksimbulva.chess.core.engine.move.generator

import ru.maksimbulva.chess.core.engine.board.Cell
import ru.maksimbulva.chess.core.engine.board.Vector2
import ru.maksimbulva.chess.core.engine.move.Move
import ru.maksimbulva.chess.core.engine.move.MoveBuilder
import ru.maksimbulva.chess.core.engine.otherPlayer
import ru.maksimbulva.chess.core.engine.position.Position

internal class RayMoveGenerator(
    private val rayDeltas: Array<Vector2>
) : IMoveGenerator {

    override fun generateMoves(fromCell: Cell, position: Position, moves: MutableList<Move>) {
        for (rayDelta in rayDeltas) {
            generateRayMoves(fromCell, position, rayDelta, moves)
        }
    }

    private fun generateRayMoves(
        fromCell: Cell,
        position: Position,
        rayDelta: Vector2,
        moves: MutableList<Move>
    ) {
        val otherPlayer = position.playerToMove.otherPlayer()
        val board = position.board
        val moveBuilder = MoveBuilder(fromCell)
        var curCell: Cell? = fromCell + rayDelta
        while (curCell != null) {
            val curMoveBuilder = moveBuilder.setToCell(curCell)
            if (board.isEmpty(curCell)) {
                moves.add(curMoveBuilder.build())
            } else {
                val obstaclePiece = board.pieceAt(curCell)
                if (obstaclePiece?.player == otherPlayer) {
                    moves.add(curMoveBuilder.setAsCapture(obstaclePiece.piece).build())
                }
                break
            }
            curCell += rayDelta
        }
    }
}

File: ChessApp/ChessEngine/Move/Generator/RayMoveGenerator.cs
using ChessEngine.Board;
using System.Collections.Generic;

namespace ChessEngine.Move.Generator
{
    internal sealed class RayMoveGenerator : IPieceMoveGenerator
    {
        private readonly Piece piece;
        private readonly MoveDelta[] rayDeltas;

        public RayMoveGenerator(Piece piece, MoveDelta[] rayDeltas)
        {
            this.piece = piece;
            this.rayDeltas = rayDeltas;
        }

        public void AppendMoves(BoardSquare originSquare, Position.Position position, List<Move> moves)
        {
            for (int i = 0; i < rayDeltas.Length; ++i)
            {
                AppendRayMoves(originSquare, position, rayDeltas[i], moves);
            }
        }

        private void AppendRayMoves(
            BoardSquare originSquare,
            Position.Position position,
            MoveDelta rayDelta,
            List<Move> moves)
        {
            var board = position.Board;
            var moveBuilder = new MoveBuilder(piece, originSquare);

            var currentSquare = originSquare;
            while (rayDelta.IsCanApplyTo(currentSquare))
            {
                currentSquare = rayDelta.GetDestSquare(currentSquare);
                if (board.IsEmpty(currentSquare))
                {
                    moves.Add(moveBuilder.SetDestSquare(currentSquare).Build());
                }
                else
                {
                    var obstacle = board.GetPieceAt(currentSquare);
                    if (obstacle.player != position.PlayerToMove)
                    {
                        moves.Add(moveBuilder
                            .SetDestSquare(currentSquare)
                            .SetCapture(obstacle.piece)
                            .Build());
                    }
                    break;
                }
            }
        }
    }
}
